vital/types/feature_track_set.h

```diff
@@ -63,15 +63,37 @@ using feature_track_set_frame_data_sptr =
 class VITAL_EXPORT feature_track_state : public track_state
 {
 public:
+  //@{
   /// Constructor
   explicit feature_track_state( frame_id_t frame,
-                                feature_sptr f = nullptr,
-                                descriptor_sptr d = nullptr)
-    : track_state( frame )
-    , feature(f)
-    , descriptor(d)
-    , inlier(false)
-  { }
+                                feature_sptr const& f = nullptr,
+                                descriptor_sptr const& d = nullptr )
+    : track_state{ frame }
+    , feature{ f }
+    , descriptor{ d }
+    , inlier{ false }
+  {}
+
+  explicit feature_track_state( frame_id_t frame,
+                                feature_sptr&& f,
+                                descriptor_sptr&& d )
+    : track_state{ frame }
+    , feature{ std::move( f ) }
+    , descriptor{ std::move( d ) }
+    , inlier{ false }
+  {}
+  //@}
+
+  /// Copy constructor
+  feature_track_state( feature_track_state const& other )
+    : track_state{ other }
+    , feature{ other.feature ? other.feature->clone() : nullptr }
```

Comment:
This may have some considerable undesired performance impact.  It may still be the right thing to do, design-wise, but may need to implement a work around to **not** deep copy the feature and descriptor.   In feature tracking, the tracks need to be frequently copied and rewritten, e.g. for merging tracks.  While the track structures are mutable, the data (feature and descriptor) are assumed to fixed and read-only after creating them.  They are also much larger, in terms of memory, than the track itself.  So by design we want to shallow copy the feature and descriptor.  However, I don't recall if the tracking algorithms are calling this code or not, so this may not be an issue.

Reply:
Yeah, I wanted to talk about that; you were out.

My use case is for object tracks, and I *absolutely* need the detection deep-copied, because I will sometimes need to use a set of tracks in multiple threads, and the detections *are* mutable. Thus, I need to be able to make a copy which has a *copy* of the detections. (Or else hope like \<redacted\> I always remember to never modify a detection in-place.)

If Marc Mutz is to be believed, all these pointers and memory indirections are actually more expensive than just making deep copies all the time with increased data locality, although personally I have a hard time buying that. If you *don't* believe that, however, I think in some ways it's a shame KWIVER adapted the "shared pointers to everything" model rather than copying Qt's implicitly shared classes, in which case this wouldn't be a problem. That, unfortunately, would be a *significant* retrofit, as well as something that would be fairly obvious to users. (I suppose I could try to implement that *just* for feature tracks... but it would still be a bigger change than this, especially since making that work with polymorphic types may require some cleverness.)

One option, of course, is for only object tracks to deep copy. I'm not sure how I would feel about that behavior difference, however.